### PR TITLE
{lang}[GCCcore/12.3.0] Rust v1.70.0

### PR DESCRIPTION
**Pull request by Micket**
(created using `eb --new-pr`)

requires:
* #18018

### CODE REVIEW

**Comment by Micket** [comment]
Test report by @Micket
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
vera-c1 - Linux Rocky Linux 8.7, x86_64, Intel Xeon Processor (Skylake), Python 3.6.8
See https://gist.github.com/Micket/831d3c2565e2ae2b95f142408056ec10 for a full test report.

**Comment by Micket** [comment]
bah, they completely changed the logic in the bootstrap.py stuff. I'll see if what i can do with the patch

**Comment by Micket** [comment]
@boegelbot please test @ generoso

**Comment by Micket** [comment]
Test report by @Micket
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
vera-c1 - Linux Rocky Linux 8.7, x86_64, Intel Xeon Processor (Skylake), Python 3.6.8
See https://gist.github.com/Micket/a2b9c83aeb873ac821e8cc0a790cd364 for a full test report.

**Comment by Micket** [comment]
@boegelbot please test @ jsc-zen2

**Comment by Micket** [comment]
Test report by @Micket
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
vera-c1 - Linux Rocky Linux 8.7, x86_64, Intel Xeon Processor (Skylake), Python 3.6.8
See https://gist.github.com/Micket/5cd748d0480bd4ebd78c8d136355a46a for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3159.skitty.os - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz (skylake_avx512), Python 3.6.8
See https://gist.github.com/boegel/0bc07bfaed95f555db33204a4eba0a59 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 3 out of 3 (1 easyconfigs in total)
fair-mastodon-c6g-2xlarge-0003 - Linux Rocky Linux 8.7, AArch64, ARM UNKNOWN (graviton2), Python 3.6.8
See https://gist.github.com/boegel/a59c235466964020626c473459c03848 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 4 out of 4 (1 easyconfigs in total)
bear-pg0105u03a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/branfosj/14aea8b14aabe3dad98ed8c21131fdc0 for a full test report.

File: easybuild/easyconfigs/r/Rust/Rust-1.70.0-GCCcore-12.3.0.eb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+dependencies = [
+    ('OpenSSL', '1.1', '', SYSTEM),
+]
+
+configopts = "--enable-extended --sysconfdir=%(installdir)s/etc "
+
+# avoid build dependency on Ninja, which requires Python,
+# since Rust is a build dependency for cryptography that is included with Python
+configopts += "--set=llvm.ninja=false "
```
Comment (branfosj) [inline easybuild/easyconfigs/r/Rust/Rust-1.70.0-GCCcore-12.3.0.eb:35]: This is no longer true. Should we switch to using Ninja?

**Review by branfosj — CHANGES_REQUESTED** [review]

File: easybuild/easyconfigs/r/Rust/Rust-1.70.0-GCCcore-12.3.0.eb (hunk 1 above)
Comment (boegel) [reply easybuild/easyconfigs/r/Rust/Rust-1.70.0-GCCcore-12.3.0.eb:35]: Yes, that makes sense to me. @Micket?

Comment (branfosj) [reply easybuild/easyconfigs/r/Rust/Rust-1.70.0-GCCcore-12.3.0.eb:35]: See #18018 for Ninja

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 10 out of 12 (1 easyconfigs in total)
node3159.skitty.os - Linux Debian GNU/Linux 11 (bullseye), x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz (skylake_avx512), Python 3.11.3
See https://gist.github.com/boegel/a7f1150f51103808138221960895e3ea for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 4 out of 4 (1 easyconfigs in total)
easybuild2.novalocal - Linux CentOS Stream 8, POWER, IBM pSeries (emulated by qemu) (power9le), Python 3.6.8
See https://gist.github.com/boegel/a5b8fd1dfbc60a7fa6e882cb5ee2d9fd for a full test report.

**Comment by Micket** [comment]
Looks like some possible issue with cmake on debian?
```
-- Could NOT find ZLIB (missing: ZLIB_LIBRARY) (found version "1.2.13")
CMake Error at Source/Modules/CMakeBuildUtilities.cmake:117 (message):
  CMAKE_USE_SYSTEM_ZLIB is ON but a zlib is not found!
```

I don't see how, since we do have zlib as a dep

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
> Looks like some possible issue with cmake on debian?
> 
> ```
> -- Could NOT find ZLIB (missing: ZLIB_LIBRARY) (found version "1.2.13")
> CMake Error at Source/Modules/CMakeBuildUtilities.cmake:117 (message):
>   CMAKE_USE_SYSTEM_ZLIB is ON but a zlib is not found!
> ```
> 
> I don't see how, since we do have zlib as a dep

It's a problem when installing CMake in an environment with an alternate sysroot, like we're doing in EESSI.
I'll install CMake using https://github.com/easybuilders/easybuild-easyblocks/pull/2248, which will fix it, sorry for the confusion.

Debian is the build container we use, but the test report basically for the EESSI build environmen, we should make EasyBuild aware of that.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
node3159.skitty.os - Linux Debian GNU/Linux 11 (bullseye), x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz (skylake_avx512), Python 3.11.3
See https://gist.github.com/boegel/06ed4211e51027845a4d4686b42c883c for a full test report.

**Comment by Micket** [comment]
Test report by @Micket
**SUCCESS**
Build succeeded for 2 out of 2 (1 easyconfigs in total)
vera-c1 - Linux Rocky Linux 8.7, x86_64, Intel Xeon Processor (Skylake), Python 3.6.8
See https://gist.github.com/Micket/e3b4d980ee49e5a68b5b67058ca5e79f for a full test report.

**Comment by boegel** [comment]
@Micket Test install in EESSI environment failed, looks like the patch needs work?
```
running: patchelf --set-interpreter /cvmfs/pilot.eessi-hpc.org/versions/2023.04/compat/linux/x86_64/lib64/ld-linux-x86-64.so.2 /tmp/vsc40023/easybuild/build/Rust/1.70.0/GCCcore-12.3.0/rustc-1.70.0-src/build/x86_64-unknown-linux-gnu/stage0/bin/cargo
Traceback (most recent call last):
  File "/tmp/vsc40023/easybuild/build/Rust/1.70.0/GCCcore-12.3.0/rustc-1.70.0-src/./x.py", line 29, in <module>
    bootstrap.main()
  File "/tmp/vsc40023/easybuild/build/Rust/1.70.0/GCCcore-12.3.0/rustc-1.70.0-src/src/bootstrap/bootstrap.py", line 966, in main
    bootstrap(args)
  File "/tmp/vsc40023/easybuild/build/Rust/1.70.0/GCCcore-12.3.0/rustc-1.70.0-src/src/bootstrap/bootstrap.py", line 930, in bootstrap
    build.download_toolchain()
  File "/tmp/vsc40023/easybuild/build/Rust/1.70.0/GCCcore-12.3.0/rustc-1.70.0-src/src/bootstrap/bootstrap.py", line 444, in download_toolchain
    self.fix_bin_or_dylib("{}/bin/cargo".format(bin_root))
  File "/tmp/vsc40023/easybuild/build/Rust/1.70.0/GCCcore-12.3.0/rustc-1.70.0-src/src/bootstrap/bootstrap.py", line 577, in fix_bin_or_dylib
    subprocess.check_output([
  File "/cvmfs/pilot.eessi-hpc.org/versions/2023.04/software/linux/x86_64/intel/skylake_avx512/software/Python/3.11.3-GCCcore-12.3.0/lib/python3.11/subprocess.py", line 466, in check_output
    return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/cvmfs/pilot.eessi-hpc.org/versions/2023.04/software/linux/x86_64/intel/skylake_avx512/software/Python/3.11.3-GCCcore-12.3.0/lib/python3.11/subprocess.py", line 548, in run
    with Popen(*popenargs, **kwargs) as process:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/cvmfs/pilot.eessi-hpc.org/versions/2023.04/software/linux/x86_64/intel/skylake_avx512/software/Python/3.11.3-GCCcore-12.3.0/lib/python3.11/subprocess.py", line 1024, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/cvmfs/pilot.eessi-hpc.org/versions/2023.04/software/linux/x86_64/intel/skylake_avx512/software/Python/3.11.3-GCCcore-12.3.0/lib/python3.11/subprocess.py", line 1917, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'nix-build'
```

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3159.skitty.os - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz (skylake_avx512), Python 3.6.8
See https://gist.github.com/boegel/cb097e4d8f4f2ba51cef472ac42611a8 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
fair-mastodon-c6g-2xlarge-0003 - Linux Rocky Linux 8.7, AArch64, ARM UNKNOWN (graviton2), Python 3.6.8
See https://gist.github.com/boegel/42368c17f61b53d34de24fd316ca8307 for a full test report.

**Comment by branfosj** [comment]
Test report by @branfosj
**SUCCESS**
Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0105u03a.bear.cluster - Linux RHEL 8.6, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/branfosj/933cca02e77aeab9daec7f4e45c4ba64 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
node3159.skitty.os - Linux Debian GNU/Linux 11 (bullseye), x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz (skylake_avx512), Python 3.11.3
See https://gist.github.com/boegel/2951433bf170169d3306e88b45141371 for a full test report.

**Comment by casparvl** [comment]
Test report by @casparvl
**FAILED**
Build succeeded for 0 out of 1 (1 easyconfigs in total)
tcn1.local.snellius.surf.nl - Linux Rocky Linux 8.7, x86_64, AMD EPYC 7H12 64-Core Processor, Python 3.6.8
See https://gist.github.com/casparvl/3c22a45b95141052bd0bd76babe8a507 for a full test report.

**Comment by boegel** [comment]
The best way to deal with the RPATH issue here is probably to create a custom easyblock for Rust, where we use `patchelf` to transform the RUNPATH section in `libLLVM*.so` into RPATH, but only when EasyBuild is configured to use RPATH.

@casparvl Thoughts on that?

**Comment by casparvl** [comment]
I had a quick look. This library doesn't seem to come with the sources, but in the logs (even with `--debug`) I don't see the command line that builds this shared library. I also checked the rpath wrappers, they definitely didn't pick up on any command line that _build_ this library (they _do_ contain command lines that _link_ this library to something else). The problem _might_ be that whatever command is used to build this library isn't intercepted by the wrappers.

I'll check if I can easily pass the `-v` flag to the build command, maybe that prints some more information...

**Comment by casparvl** [comment]
From the `./x.py build -v` output:

```
...
download https://ci-artifacts.rust-lang.org/rustc-builds/90c541806f23a127002de5b4038be731ba1458ca/rust-dev-1.70.0-x86_64-unknown-linux-gnu.tar.xz
downloading https://ci-artifacts.rust-lang.org/rustc-builds/90c541806f23a127002de5b4038be731ba1458ca/rust-dev-1.70.0-x86_64-unknown-linux-gnu.tar.xz
running: "curl" "-#" "-y" "30" "-Y" "10" "--connect-timeout" "30" "--retry" "3" "-Sf" "https://ci-artifacts.rust-lang.org/rustc-builds/90c541806f23a127002de5b4038be731ba1458ca/rust-dev-1.70.0-x86_64-unknown-linux-gnu.tar.xz"
extracting /gpfs/nvme1/1/casparl/ebbuildpath/Rust/1.70.0/GCCcore-12.3.0/rustc-1.70.0-src/build/cache/llvm-90c541806f23a127002de5b4038be731ba1458ca-false/rust-dev-1.70.0-x86_64-unknown-linux-gnu.tar.xz to /gpfs/nvme1/1/casparl/ebbuildpath/Rust/1.70.0/GCCcore-12.3.0/rustc-1.70.0-src/build/x86_64-unknown-linux-gnu/ci-llvm
... 
extracting rust-dev-1.70.0-x86_64-unknown-linux-gnu/rust-dev/lib/libLLVM-16-rust-1.70.0-stable.so to /gpfs/nvme1/1/casparl/ebbuildpath/Rust/1.70.0/GCCcore-12.3.0/rustc-1.70.0-src/build/x86_64-unknown-linux-gnu/ci-llvm
...
``` 
That sounds like it is distributed as a binary, and not compiled...?

**Comment by casparvl** [comment]
Ok, I'm now 99% sure this is just installed as a binary:
```
$ cd /gpfs/nvme1/1/casparl/ebbuildpath/Rust/1.70.0/GCCcore-12.3.0/rustc-1.70.0-src/build/cache/llvm-90c541806f23a127002de5b4038be731ba1458ca-false
[casparl@tcn37 llvm-90c541806f23a127002de5b4038be731ba1458ca-false]$ ls
rust-dev-1.70.0-x86_64-unknown-linux-gnu.tar.xz
[casparl@tcn37 llvm-90c541806f23a127002de5b4038be731ba1458ca-false]$ tar -xf rust-dev-1.70.0-x86_64-unknown-linux-gnu.tar.xz
[casparl@tcn37 llvm-90c541806f23a127002de5b4038be731ba1458ca-false]$ cd rust-dev-1.70.0-x86_64-unknown-linux-gnu/
[casparl@tcn37 rust-dev-1.70.0-x86_64-unknown-linux-gnu]$ find -name libLLVM*
./rust-dev/lib/libLLVM-16-rust-1.70.0-stable.so
[casparl@tcn37 rust-dev-1.70.0-x86_64-unknown-linux-gnu]$ sha256sum rust-dev/lib/libLLVM-16-rust-1.70.0-stable.so
1970a0cb1b44494f6c6095341947a84bccad37654c34a724a64c98e5a4ed024e  rust-dev/lib/libLLVM-16-rust-1.70.0-stable.so
```
And, on the installed library, I get the same checksum:
``` 
$ sha256sum /scratch-nvme/1/casparl/generic/software/Rust/1.70.0-GCCcore-12.3.0/lib/libLLVM-16-rust-1.70.0-stable.so
1970a0cb1b44494f6c6095341947a84bccad37654c34a724a64c98e5a4ed024e  /scratch-nvme/1/casparl/generic/software/Rust/1.70.0-GCCcore-12.3.0/lib/libLLVM-16-rust-1.70.0-stable.so
```

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Going in, thanks @Micket!

**Pull request merged**